contributors/design-proposals/service-injection-policy.md

```diff
@@ -0,0 +1,404 @@
+
+## Motivation
+
+Consuming a service involves more than just connectivity.  In addition to
+coordinates to reach the service, credentials and non-secret configuration
+parameters are typically needed to use the service.  The primitives for this
+already exist, but a gap exists where loose coupling is desired: it should be
+possible to inject pods with the information they need to use a service on a
+service-by-service basis, without the pod authors having to incorporate the
+information into every pod spec where it is needed.
+
+## Constraints and Assumptions
+
+1.  New mechanisms must be made to work with controllers such as deployments and
+    replicasets that create pods. Existing controllers that create pods should
+    recreate their pods when a new Service Injection Policy is added that would
+    effect them.
+
+## Use Cases
+
+1.  As a user, I want to be able to describe a way that pods should be injected
+    with the information to consume a particular service in a loosely-coupled
+    way, so that I can concisely model the information about how the service
+    should be consumed without altering every consuming pod spec.
+
+### Loose coupling between services and their consumers
+
+Using a Service Injection Policy allows pod authors to not have to explicitly
+set information for every pod. The information will be injected based off the
+policy's `Selector`.
+
+## Proposed Changes
+
+### ServiceInjectionPolicy API object
+
+```go
+type ServiceInjectionPolicy struct {
+    unversioned.TypeMeta
+    ObjectMeta
+
+    Spec ServiceInjectionPolicySpec
+}
+
+type ServiceInjectionPolicySpec struct {
+    // Selector defines which Pods should be modified (required, immutable).
+    Selector      *unversioned.LabelSelector
+    // Env defines the collection of EnvVar to inject (optional, mutable).
+    Env           []EnvVar
+    // EnvFrom defines the collection of EnvFromSource to inject (optional, mutable).
+    EnvFrom       []EnvFromSource
+    // Volumes defines the collection of Volume to inject (optional, mutable).
+    Volumes       []Volume
+    // VolumeMounts defines the collection of VolumeMount to inject (optional, mutable).
+    VolumeMounts  []VolumeMount
+}
+```
+
+### AdmissionControl Plugin: ServiceInjectionPolicy
+
+The **ServiceInjectionPolicy** plugin introspects all incoming pod creation
+requests and injects the pod based off a `Selector` with the desired
+attributes.
+
+The order of precedence for applying multiple `ServiceInjectionPolicy` specs is
+from oldest to newest.
+
+#### Behavior
+
+This will modify the containers in the pod spec. The supported changes to
+`Env`, `EnvFrom`, and `VolumeMounts` apply to the container spec for
+all containers in the pod with the specified matching `Selector`. The
+changes to `Volumes` apply to the pod spec for all pods matching `Selector`.
+
+*Why modify all containers in a pod?*
+
+Currently there is no concept of labels on specific containers in a pod which
```

Comment:
If container labels were added, we would have to think carefully about how to take advantage of them without breaking backward compatibility.  We couldn't just reuse the existing labels and gate on those, because it would change the api behavior for containers without labels.  In theory we could do something weird like - inject in containers with no-labels or containers with matching labels, but not containers with labels that don't match.  I think that logic is a bit counter-intuitive for new users though.

Summary: In the future, even if container labels are added, we will still need to add another field to the SIP to specify the container labels separately from the Pod labels.